### PR TITLE
[Structural] Adding Orthotropic damage constitutive law

### PR DESCRIPTION
**Pull request by AlejandroCornejo**
In this PR I'm adding an orthotropic damage law according the cervera et al in "Appraisement of planar, bending and twisting cracks in 3D with isotropic and orthotropic damage models" Int J Fract (2018) 210:45–79

### CODE REVIEW

**Comment by loumalouomega** [comment]
The commits are not very descriptive :P

**Comment by AlejandroCornejo** [comment]
> The commits are not very descriptive :P

yeah sorry... I was in a hurry... I'll avoid that if you actually use them

**Comment by philbucher** [comment]
how do you specify the material-orientation?

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_orthotropic_damage.cpp [hunk 1]
```diff
@@ -0,0 +1,725 @@
...
+        } // Damages computed
+
+        // We compute the secant tensor in principal axis
+        Matrix secant_tensor(VoigtSize, VoigtSize);
+        noalias(secant_tensor) = ZeroMatrix(VoigtSize, VoigtSize);
+        this->CalculateSecantTensor(secant_tensor, rValues, damages);
+
+        // Now we recover the original axis system
+        Matrix auxiliar(VoigtSize,VoigtSize);
+        auxiliar = prod(secant_tensor, rotation_matrix);
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_orthotropic_damage.cpp]: Noalias

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_orthotropic_damage.cpp [hunk 2]
```diff
@@ -0,0 +1,725 @@
...
+
+template <class TConstLawIntegratorType>
+void GenericSmallStrainOrthotropicDamage<TConstLawIntegratorType>::FinalizeSolutionStep(
+    const Properties& rMaterialProperties,
+    const GeometryType &rElementGeometry,
+    const Vector& rShapeFunctionsValues,
+    const ProcessInfo& rCurrentProcessInfo
+    )
+{
+    // Deprecated
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_orthotropic_damage.cpp]: Remove it if not used (and deprecated)

**Review by loumalouomega — COMMENTED** [review]
I will continue reviewing later

**Comment by AlejandroCornejo** [comment]
> how do you specify the material-orientation?

It is not exactly that kind of orthotropy. This model is used for isotropic materials but with directional damage (see the reference). This means that the material is going to be damaged on each principal direction independently.

**Comment by philbucher** [comment]
> > how do you specify the material-orientation?
> 
> It is not exactly that kind of orthotropy. This model is used for isotropic materials but with directional damage (see the reference). This means that the material is going to be damaged on each principal direction independently.

ok thx for the info

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_orthotropic_damage.cpp [hunk 3]
```diff
@@ -0,0 +1,709 @@
...
+        int i=0,j=0,k=0;
+        if (a >= b && b >= c) {i=0; j=1; k=2;}
+        else if (a >= c && c >= b) {i=0; j=2; k=1;}
+        else if (b >= a && a >= c) {i=1; j=0; k=2;}
+        else if (b >= c && c >= a) {i=1; j=2; k=2;}
+        else if (c >= a && a >= b) {i=2; j=0; k=1;}
+        else if (c >= b && b >= a) {i=2; j=1; k=0;}
+        else KRATOS_ERROR << "Problem with the principal stresses..." << std::endl;
+
+        Matrix copy(VoigtSize,VoigtSize);
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_orthotropic_damage.cpp]:
```suggestion
        Matrix copy(rEigenVectorsMatrix);
```

Easier

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_orthotropic_damage.cpp [hunk 4]
```diff
@@ -0,0 +1,709 @@
...
+        copy.resize(VoigtSize,VoigtSize);
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_orthotropic_damage.cpp]:
```suggestion
```

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_orthotropic_damage.cpp [hunk 5]
```diff
@@ -0,0 +1,709 @@
...
+        noalias(copy) = rEigenVectorsMatrix;
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_orthotropic_damage.cpp]:
```suggestion
```

**Review by loumalouomega — APPROVED** [review]
Minor comments, and where is the test? 👀

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Minor comments, and where is the test? 👀

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_orthotropic_damage.cpp (hunk 3 above)
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_orthotropic_damage.cpp]: this can be done? OMG

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_orthotropic_damage.cpp]: https://www.geeksforgeeks.org/copy-constructor-in-cpp/

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_orthotropic_damage.cpp]: It is called copy constructor

imagen

Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_orthotropic_damage.cpp]: image
I knew about the copy but not with the entity Matrix :)

**Comment by AlejandroCornejo** [comment]
> Minor comments, and where is the test? 👀

You're right, tomorrow I'll add one

**Comment by AlejandroCornejo** [comment]
comments adressed

**Comment by loumalouomega** [comment]
BTW the reason why the commits must be descriptive is because when you track back on the commit history you have an idea of what's going on

**Review by loumalouomega — APPROVED** [review]
👌

**Pull request merged**